platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/impl/DefaultExecutionProblemHandler.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.impl;
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.problems.Severity;
+import org.gradle.api.problems.internal.InternalProblem;
+import org.gradle.api.problems.internal.InternalProblemReporter;
+import org.gradle.api.problems.internal.InternalProblems;
+import org.gradle.internal.execution.ExecutionProblemHandler;
+import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.execution.WorkValidationContext;
+import org.gradle.internal.execution.WorkValidationException;
+import org.gradle.internal.execution.steps.ValidateStep;
+import org.gradle.internal.reflect.validation.TypeValidationProblemRenderer;
+import org.gradle.internal.vfs.VirtualFileSystem;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+import static org.gradle.api.problems.Severity.ERROR;
+import static org.gradle.api.problems.Severity.WARNING;
+
+public class DefaultExecutionProblemHandler implements ExecutionProblemHandler {
```

Comment:
💭 This has been moved out of `ValidateStep` so other steps can also report warnings and fail on validation errors.